codex/erasure/asyncbackend.nim

```diff
@@ -0,0 +1,62 @@
+import pkg/taskpools
+
+import pkg/chronos
+import pkg/chronos/threadsync
+
+import pkg/chronicles
+import pkg/questionable
+import pkg/questionable/results
+
+import ./backend
+
+type
+  EncoderArgs* = object
+    backend*: ptr EncoderBackend
+    data*: ptr seq[seq[byte]]
+    parity*: ptr seq[seq[byte]]
+
+  ThreadArgs = object
+    signal: ThreadSignalPtr
+    encoderArgs: EncoderArgs
+
+  ThreadResult = Result[void, cstring]
+
+proc encode(args: ThreadArgs): ThreadResult =
+
+  try:
+    let res = args.encoderArgs.backend[].encode((args.encoderArgs.data[]), (args.encoderArgs.parity[]))
+    return res
+  except Exception as exception:
+    return err(exception.msg.cstring)
+  finally:
+    let fireRes = args.signal.fireSync()
+    if fireRes.isErr:
+      echo "Error firing signal: " & fireRes.error
+      # return err("Error firing signal: " & fireRes.error)
+
+proc proxy(tp: Taskpool, args: ThreadArgs): Flowvar[ThreadResult] =
+  tp.spawn encode(args)
+
+proc asyncEncode*(
+  tp: Taskpool,
+  args: EncoderArgs
+): Future[?!void] {.async.} =
+  let signalRes = ThreadSignalPtr.new()
+
+  if signalRes.isErr:
+    return failure(signalRes.error)
+  else:
+    let signal = signalRes.value
+
+    var args = ThreadArgs(signal: signal, encoderArgs: args)
+
+    let handle = proxy(tp, args)
+
+    await wait(signal) # TODO consider adding timeout
+
+    let res = sync handle # should be completed
```

Comment:
Doing this will turn the current main thread into a worker with taskpools. It assumes that if you're doing a sync you're waiting and can do work. 

What you will want to do is to use `tryComplete(handle, res)` from https://github.com/status-im/nim-taskpools/blob/d4c43137c0590cb47f893a66ca8cb027fa6c217e/taskpools/flowvars.nim#L56C10-L56C21

Reply:
If you're curious, here's where that occurs: https://github.com/status-im/nim-taskpools/blob/d4c43137c0590cb47f893a66ca8cb027fa6c217e/taskpools/taskpools.nim#L247

Reply:
This is correct, I believe we might want to poll with`tryComplete` here.

Reply:
Right, I will change it to `tryComplete` and return `failure(...)` if it happens to not be completed - which it shouldn't.